plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+							}
+						}
+					}
+				}
+			}
+			#ifdef SA_DEBUG
+				// report FFT processing speed
+				start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start_time;
+				std::cout << "Processed " << m_framesFilledUp << " samples in " << start_time / 1000000.0 << " ms" << std::endl;
+			#endif
+
+			// clean up before checking for more data from input buffer
+			m_framesFilledUp = 0;
+		}
+	}
+}
+
+
+// Produce a spectrogram pixel from normalized spectrum data.
+// Values over 1.0 will cause the color components to overflow: this is left
+// intentionally untreated as it clearly indicates which frequency is clipping.
+// Gamma correction is applied to make small values more visible and to make
+// a linear gradient actually appear roughly linear. The correction should be
+// around 0.42 to 0.45 for sRGB displays (or lower for bigger visibility boost).
+QRgb SaProcessor::makePixel(float left, float right, float gamma_correction)
+{
+	if (m_controls->m_stereoModel.value())
+	{
+		float ampL = pow(left, gamma_correction);
+		float ampR = pow(right, gamma_correction);
+		return qRgb(m_controls->m_colorL.red() * ampL + m_controls->m_colorR.red() * ampR,
+					m_controls->m_colorL.green() * ampL + m_controls->m_colorR.green() * ampR,
+					m_controls->m_colorL.blue() * ampL + m_controls->m_colorR.blue() * ampR);
+	}
+	else
+	{
+		float ampL = pow(left, gamma_correction);
+		// make mono color brighter to compensate for the fact it is not summed
+		return qRgb(m_controls->m_colorMono.lighter().red() * ampL,
+					m_controls->m_colorMono.lighter().green() * ampL,
+					m_controls->m_colorMono.lighter().blue() * ampL);
+	}
+}
+
+
+// Get sample rate value that is valid for currently stored results.
+int SaProcessor::getSampleRate() const
+{
+	return m_sampleRate;
+}
+
+
+// Inform the processor whether any display widgets actually need it.
+void SaProcessor::setSpectrumActive(bool active)
+{
+	m_spectrumActive = active;
+}
+
+void SaProcessor::setWaterfallActive(bool active)
+{
+	m_waterfallActive = active;
+}
+
+
+// Reallocate data buffers according to newly set block size.
+void SaProcessor::reallocateBuffers()
+{
+	// The SaProcessor destructor is called before SaControlsDialog destructor.
+	// This causes problems, sice ComboBoxModel emits dataChanged() as it is
+	// destroyed, triggering reallocation in processor that is already gone --
+	// a good recipe for segfault. This check explicitly prevents it.
+	if (m_destroyed) {return;}
+
+	int new_size_index = m_controls->m_blockSizeModel.value();
+	int new_in_size, new_fft_size;
+	int new_bins;
+
+	// get new block sizes and bin count based on selected index
+	if (new_size_index < FFT_BLOCK_SIZES.size())
+	{
+		new_in_size = FFT_BLOCK_SIZES[new_size_index];
+	}
+	else
+	{
+		new_in_size = FFT_BLOCK_SIZES.back();
+	}
+	if (new_size_index + m_zeroPadFactor < FFT_BLOCK_SIZES.size())
+	{
+		new_fft_size = FFT_BLOCK_SIZES[new_size_index + m_zeroPadFactor];
```

Comment:
It's just a safe value -- the FFT_BLOCK_SIZES will always contain the zero element (otherwise it does not make any sense), but depending on the m_zeroPadFactor, it may not contain element [0 + m_zeroPadFactor] and dealing with this would just make the contructor more complicated (with no benefit, since the first thing that happens when controls are "connected" is reallocation to the default block size).

Reply:
Oh, ofc, now I see it. Resolving...